### ISSUE TITLE
kubelet built from master panics: "invalid memory address or nil pointer dereference"

### ISSUE THREAD

**Issue by luxas** [kubernetes/kubernetes#52242]
/kind bug

**What happened**:

I ran kubelet from HEAD and tried to make a normal cluster using kubeadm. I saw errors like these in the kubelet's log:

(I think [this line](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/validation/validation.go#L996) is the cause)

```
syys 10 13:12:00 THEGOPHER kubeletHEAD[22300]: E0910 13:12:00.450692   22300 runtime.go:66] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
syys 10 13:12:00 THEGOPHER kubeletHEAD[22300]: vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:72
syys 10 13:12:00 THEGOPHER kubeletHEAD[22300]: vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
syys 10 13:12:00 THEGOPHER kubeletHEAD[22300]: vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
syys 10 13:12:00 THEGOPHER kubeletHEAD[22300]: /root/.cache/bazel/_bazel_root/e9f728bbd90b3fba632eb31b20e1dacd/external/go1_8_3_linux_amd64/src/runtime/asm_amd64.s:514
syys 10 13:12:00 THEGOPHER kubeletHEAD[22300]: /root/.cache/bazel/_bazel_root/e9f728bbd90b3fba632eb31b20e1dacd/external/go1_8_3_linux_amd64/src/runtime/panic.go:489
syys 10 13:12:00 THEGOPHER kubeletHEAD[22300]: /root/.cache/bazel/_bazel_root/e9f728bbd90b3fba632eb31b20e1dacd/external/go1_8_3_linux_amd64/src/runtime/panic.go:63
syys 10 13:12:00 THEGOPHER kubeletHEAD[22300]: /root/.cache/bazel/_bazel_root/e9f728bbd90b3fba632eb31b20e1dacd/external/go1_8_3_linux_amd64/src/runtime/signal_unix.go:290
syys 10 13:12:00 THEGOPHER kubeletHEAD[22300]: pkg/api/validation/validation.go:996
syys 10 13:12:00 THEGOPHER kubeletHEAD[22300]: pkg/api/validation/validation.go:631
syys 10 13:12:00 THEGOPHER kubeletHEAD[22300]: pkg/api/validation/validation.go:406
syys 10 13:12:00 THEGOPHER kubeletHEAD[22300]: pkg/api/validation/validation.go:364
syys 10 13:12:00 THEGOPHER kubeletHEAD[22300]: pkg/api/validation/validation.go:2482
syys 10 13:12:00 THEGOPHER kubeletHEAD[22300]: pkg/api/validation/validation.go:2448
syys 10 13:12:00 THEGOPHER kubeletHEAD[22300]: pkg/kubelet/config/config.go:334
syys 10 13:12:00 THEGOPHER kubeletHEAD[22300]: pkg/kubelet/config/config.go:232
syys 10 13:12:00 THEGOPHER kubeletHEAD[22300]: pkg/kubelet/config/config.go:287
syys 10 13:12:00 THEGOPHER kubeletHEAD[22300]: pkg/kubelet/config/config.go:159
syys 10 13:12:00 THEGOPHER kubeletHEAD[22300]: pkg/util/config/config.go:83
syys 10 13:12:00 THEGOPHER kubeletHEAD[22300]: pkg/util/config/config.go:77
syys 10 13:12:00 THEGOPHER kubeletHEAD[22300]: vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133
syys 10 13:12:00 THEGOPHER kubeletHEAD[22300]: vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134
syys 10 13:12:00 THEGOPHER kubeletHEAD[22300]: vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
syys 10 13:12:00 THEGOPHER kubeletHEAD[22300]: /root/.cache/bazel/_bazel_root/e9f728bbd90b3fba632eb31b20e1dacd/external/go1_8_3_linux_amd64/src/runtime/asm_amd64.s:2197
```

**What you expected to happen**:

No panic

**How to reproduce it (as minimally and precisely as possible)**:

1. Install kubeadm the normal way
2. Download kubelet, kubeadm from CI HEAD and `kubeadm init` a cluster like normal
3. kubelet crashes

Seems related to https://github.com/kubernetes/features/issues/376
@kubernetes/sig-storage-bugs @kubernetes/sig-node-bugs

**Comment by luxas** [comment]
cc @dixudx @thockin 

I think the only thing we have to do here is to check for nil before trying to dereference the pointer

Adding the v1.8 milestone to this bug to make we track it in a release-blocking sense, this is probably a release blocker. cc @kubernetes/kubernetes-release-managers

**Comment by dixudx** [comment]
@luxas Since we've already set default value for `HostPathType`, the value is a non-nil pointer.

https://github.com/kubernetes/kubernetes/blob/e821e531db24ddd4897a166b67af8c863d93edc2/pkg/api/v1/defaults.go#L386-L391

No obvious errors found in [kubeadm build-log.txt](https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-kubeadm-gce-1-7-on-1-8/4/build-log.txt) are related to `HostPathType`.

**Comment by luxas** [comment]
Yeah, but somehow it seems to be nil here... otherwise it wouldn't have paniced

**Comment by thockin** [comment]
Before we merge, we need to understand how it got to be nil.  It's supposed to be defaulted.

**Comment by luxas** [comment]
@thockin Absolutely. @dixudx has a prototype of how it could be nil now in https://github.com/kubernetes/kubernetes/pull/52246, let's verify if it fixes the issue...

**Comment by yujuhong** [comment]
> Before we merge, we need to understand how it got to be nil. It's supposed to be defaulted.

I think volume manager reconstructs the specs from reading the directories on the disk and these specs do not have the defaults applied.
@kubernetes/sig-storage-bugs would know better

**Comment by liggitt** [comment]
> I think volume manager reconstructs the specs from reading the directories on the disk and these specs do not have the defaults applied.

Using api machinery validation without accompanying api machinery decoding/defaulting is dangerous

**Comment by dchen1107** [comment]
Agreed with both @yujuhong and @liggitt above. @saad-ali can you assign this to the owner of the volume manager in sig-storage? 

cc/ @kubernetes/sig-storage-bugs

**Comment by dixudx** [comment]
> I think volume manager reconstructs the specs from reading the directories on the disk and these specs do not have the defaults applied.

@dchen1107 The root cause is apiserver version < kubelet version, as I stated in [discussion](https://github.com/kubernetes/kubernetes/pull/52246#issuecomment-329361077).

**Comment by liggitt** [comment]
Even with an old api server, it concerns me that there is apparently a code path that causes the kubelet to run validation on internal objects that did not go through a defaulting pass as part of being converted from v1 objects to internal objects.

**Comment by saad-ali** [comment]
This issue is being discussed on https://github.com/kubernetes/kubernetes/pull/52246, see the latest on that PR. At the moment, it appears to be caused by kubeadm using a newer version of the k8s client then then the server. Awaiting word from @jessicaochen and @luxas. If that is the case, we should be able to close this bug. If not, we can merge https://github.com/kubernetes/kubernetes/pull/52246 as a temporary mitigation to unblock release.

**Comment by roberthbailey** [comment]
Also, your comment doesn't address @liggitt's concern above.

**Comment by saad-ali** [comment]
> Do you mean kubelet? We explicitly support clients being ahead of the server by a minor version.

Yes, referencing this discussion (https://github.com/kubernetes/kubernetes/pull/52246#issuecomment-329903096):

> thockin commented 4 days ago:
> In general we don't support moving the kubelet to be newer than the apiserver. If we do that, then yes, we need to nil-check EVERY optional field.

---------

> Also, your comment doesn't address @liggitt's concern above.
>> liggitt commented 17 hours ago
>> Even with an old api server, it concerns me that there is apparently a code path that causes the kubelet to run validation on internal objects that did not go through a defaulting pass as part of being converted from v1 objects to internal objects.

I don't think that is the case. See @msau42's comment here: https://github.com/kubernetes/kubernetes/pull/52246#issuecomment-328729460. Based on that discussion, it looks like volume reconstruction was a red herring. In general volume reconstruction creates internal objects that are only used internally, they are not run through validation.

**Comment by saad-ali** [comment]
Still awaiting response on https://github.com/kubernetes/kubernetes/pull/52246 CC @luxas @jessicaochen

**Comment by saad-ali** [comment]
> I don't think that is the case. See @msau42's comment here: #52246 (comment). Based on that discussion, it looks like volume reconstruction was a red herring. In general volume reconstruction creates internal objects that are only used internally, they are not run through validation.

Volume reconstruction is not entirely guilt free here: it constructs an internal object, does not run the object through validation or defaulting logic, so if a new field is introduced (like hostpath type in this case), the default value is not set resulting in issues like this one. To handle this case gracefully I'm approving PR https://github.com/kubernetes/kubernetes/pull/52246. For Q4 (1.9) we will follow up on this https://github.com/kubernetes/kubernetes/issues/52820 and see what we can do to handle this more gracefully. In general, volume reconstruction is a thorny bit of code and deserves a thorough redesign.

**Comment by jingxu97** [comment]
I have filed an issue related to internal volume reconstruction process during kublet restarts. The new proposed approach is very different from the current design. Will work on it in detail for 1.9 
https://github.com/kubernetes/kubernetes/issues/52683

**Comment by qinzhao168** [comment]
I got the name bug

this is log

Oct 24 14:45:27 node-23-40 kubelet[14309]: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:139
Oct 24 14:45:27 node-23-40 kubelet[14309]: /usr/local/go/src/runtime/asm_amd64.s:2361
Oct 24 14:45:27 node-23-40 kubelet[14309]: E1024 14:45:27.235639   14309 runtime.go:66] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
Oct 24 14:45:27 node-23-40 kubelet[14309]: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:72
Oct 24 14:45:27 node-23-40 kubelet[14309]: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
Oct 24 14:45:27 node-23-40 kubelet[14309]: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
Oct 24 14:45:27 node-23-40 kubelet[14309]: /usr/local/go/src/runtime/asm_amd64.s:573
Oct 24 14:45:27 node-23-40 kubelet[14309]: /usr/local/go/src/runtime/panic.go:502
Oct 24 14:45:27 node-23-40 kubelet[14309]: /usr/local/go/src/runtime/panic.go:63
Oct 24 14:45:27 node-23-40 kubelet[14309]: /usr/local/go/src/runtime/signal_unix.go:388
Oct 24 14:45:27 node-23-40 kubelet[14309]: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/volume/flexvolume/plugin.go:85
Oct 24 14:45:27 node-23-40 kubelet[14309]: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/volume/flexvolume/probe.go:130
Oct 24 14:45:27 node-23-40 kubelet[14309]: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/volume/flexvolume/probe.go:88
Oct 24 14:45:27 node-23-40 kubelet[14309]: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/volume/flexvolume/probe.go:88
Oct 24 14:45:27 node-23-40 kubelet[14309]: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/volume/flexvolume/probe.go:77
Oct 24 14:45:27 node-23-40 kubelet[14309]: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/volume/plugins.go:593
Oct 24 14:45:27 node-23-40 kubelet[14309]: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/volume/plugins.go:540
Oct 24 14:45:27 node-23-40 kubelet[14309]: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/volumemanager/cache/desired_state_of_world.go:192

**Comment by msau42** [comment]
@qinzhao168 what version did you see the panic on?

**Comment by qinzhao168** [comment]
@msau42 

 kubectl version 
Client Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.2", GitCommit:"bb9ffb1654d4a729bb4cec18ff088eacc153c239", GitTreeState:"clean", BuildDate:"2018-09-19T07:01:43Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"11+", GitVersion:"v1.11.0-168+f47446a730ca03", GitCommit:"f47446a730ca037473fb3bf0c5abeea648c1ac12", GitTreeState:"clean", BuildDate:"2018-08-25T21:05:52Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}

**Comment by msau42** [comment]
What flex driver are you using? It looks like it's not returning any capabilities.

Here's the panic line:
https://sourcegraph.com/github.com/kubernetes/kubernetes@release-1.11/-/blob/pkg/volume/flexvolume/plugin.go#L85

cc @chakri-nelluri should we fallback to default capabilities if the driver doesn't return any?

**Comment by msau42** [comment]
This stacktrace is unrelated to the original failure. Can you open a new issue for this?

**Comment by chakri-nelluri** [comment]
@msau42 We already default it, unless the plugin overrides it with nil: https://sourcegraph.com/github.com/kubernetes/kubernetes@release-1.11/-/blob/pkg/volume/flexvolume/driver-call.go#L248:1
